pkgs/tools/misc/file/default.nix

```diff
@@ -7,31 +7,24 @@
 
 stdenv.mkDerivation rec {
   pname = "file";
-  version = "5.44";
+  version = "5.45";
 
   src = fetchurl {
     urls = [
       "https://astron.com/pub/file/${pname}-${version}.tar.gz"
       "https://distfiles.macports.org/file/${pname}-${version}.tar.gz"
     ];
-    sha256 = "sha256-N1HH+6jbyDHLjXzIr/IQNUWbjOUVXviwiAon0ChHXzs=";
+    hash = "sha256-/Jf1ECm7DiyfTjv/79r2ePDgOe6HK53lwAKm0Jx4TYI=";
   };
 
-  outputs = [ "out" "dev" "man" ];
-
   patches = [
-    # Backport fix to identification for pyzip files.
-    # Needed for strip-nondeterminism.
-    # https://salsa.debian.org/reproducible-builds/strip-nondeterminism/-/issues/20
-    ./pyzip.patch
-
-    # Backport fix for --uncompress always detecting contents as "empty"
-    (fetchurl {
-      url = "https://gitweb.gentoo.org/repo/gentoo.git/plain/sys-apps/file/files/file-5.44-decompress-empty.patch?h=dfc57da515a2aaf085bea68267cc727f1bfaa691";
-      hash = "sha256-fUzRQAlLWczBmR5iA1Gk66mHjP40MJcMdgCtm2+u1SQ=";
-    })
+    # Upstream patch to fix 32-bit tests.
+    # WIll be included in 5.46+ releases.
+    ./32-bit-time_t.patch
```

Comment:
Why not `fetchpatch`?

Reply:
`fetchpatch` has `file` in it's dependencies. As a result we can't use it for most of packages in `stdenv` closure:

```diff
--- a/pkgs/tools/misc/file/default.nix
+++ b/pkgs/tools/misc/file/default.nix
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchurl, file, zlib, libgnurx }:
+{ lib, stdenv, fetchurl, fetchpatch, file, zlib, libgnurx }:

 # Note: this package is used for bootstrapping fetchurl, and thus
 # cannot use fetchpatch! All mutable patches (generated by GitHub or
@@ -22,7 +22,11 @@ stdenv.mkDerivation rec {
   patches = [
     # Upstream patch to fix 32-bit tests.
     # WIll be included in 5.46+ releases.
-    ./32-bit-time_t.patch
+    # ./32-bit-time_t.patch
+    (fetchpatch {
+      url = "https://github.com/file/file/commit/218fdf813fd5ccecbb8887a1b62509cd1c6dd3a1.patch";
+      hash = "sha256-/Jf1ECm1DiyfTjv/79r2ePDgOe6HK53lwAKm0Jx4TYI=";
+    })
   ];

   strictDeps = true;
```

```
$ nix build -f. file
error:
       … while evaluating a branch condition

         at /mnt/archive/big/git/nixpkgs-master/pkgs/stdenv/booter.nix:99:7:

           98|     thisStage =
           99|       if args.__raw or false
             |       ^
          100|       then args'

       … in the right operand of the update (//) operator

         at /mnt/archive/big/git/nixpkgs-master/pkgs/stdenv/booter.nix:84:7:

           83|       { allowCustomOverrides = index == 1; }
           84|       // (stageFun prevStage))
             |       ^
           85|     (lib.lists.reverseList stageFuns);

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: function 'anonymous lambda' called with unexpected argument 'postFetch'

       at /mnt/archive/big/git/nixpkgs-master/pkgs/build-support/fetchurl/boot.nix:5:1:

            4|
            5| { url ? builtins.head urls
             | ^
            6| , urls ? []
```

Reply:
And what about `fetchurl`? That was used before to fetch the gentoo patch.

Reply:
`fetchurl` would be fine if we fetched a file with known contents upfront. The gentoo patch was a static file in their repository.

Unfortunately links like https://github.com/file/file/commit/218fdf813fd5ccecbb8887a1b62509cd1c6dd3a1.patch are not guaranteed to be stable from one to another `github` release and thus `fetchurl` will fail it's hash from time to time (when metadata around the patch changes). `fetchpatch` is usually a solution for cases like those as it strips metadata first and then calculates the hash). But `fetchpatch` is not available for `stdenv` closure.

Reply:
@trofi is correct.  This is an unfortunate limitation of nixpkgs (and nix, [except very bleeding-edge versions](https://github.com/NixOS/nixpkgs/pull/188587)) right now.

Reply:
> Unfortunately links like [file/file@`218fdf8`.patch](https://github.com/file/file/commit/218fdf813fd5ccecbb8887a1b62509cd1c6dd3a1.patch) are not guaranteed to be stable from one to another `github` release

Interesting.. Any idea why is that? What about https://github.com/file/file/commit/218fdf813fd5ccecbb8887a1b62509cd1c6dd3a1.diff ? Seems without metadata to me.

Reply:
> Interesting.. Any idea why is that?

Good question. I would guess that long-tern stability is not their priority (just like `git diff` or `patch` slightly change their output from release to release).

> What about https://github.com/file/file/commit/218fdf813fd5ccecbb8887a1b62509cd1c6dd3a1.diff ? Seems without metadata to me.

Yeah, those look better. It still has a chance break in future around `index 2e0494d2f..78f574ea1 100644` and use more than 9 digits of abbreviation. I guess digit count is dynamically increasing with repository's commit count.

For example `linux` uses 14-digit abbreviation right now: https://github.com/torvalds/linux/commit/024ff300db33968c133435a146d51ac22db27374.diff
`index 1fbda2f94184e3..b21335e6a210e6 100644`

And `gcc` uses 12 digits: https://github.com/gcc-mirror/gcc/commit/5b42ee2cded7c2173ea0e9f9e08dd1d9b226ed3e.diff `index 181df09e5879..67ac03558b6a 100644`

Reply:
> Interesting.. Any idea why is that? What about https://github.com/file/file/commit/218fdf813fd5ccecbb8887a1b62509cd1c6dd3a1.diff ? Seems without metadata to me.

@doronbehar, see the comment at the top of `fetchpatch/default.nix`.